03_cross_cancer_classification/plot_holdout_pca.ipynb

```diff
@@ -0,0 +1,385 @@
+      "text/plain": [
+       "<Figure size 1296x720 with 8 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "gene = 'BRAF'\n",
+    "cancer_type = 'THCA'\n",
+    "projection_method = 'umap'\n",
+    "remove_percents = [0.1, 0.5, 0.75, 0.9]\n",
+    "sns.set({'figure.figsize': (18, 10)})\n",
+    "\n",
+    "fig, axarr = plt.subplots(2, len(remove_percents))\n",
+    "\n",
+    "z_label = '{} wild-type'.format(gene)\n",
+    "nz_label = '{} mutant'.format(gene)\n",
+    "x_label = '{}1'.format(projection_method.upper())\n",
+    "y_label = '{}2'.format(projection_method.upper())\n",
+    "\n",
+    "for ix, remove_percent in enumerate(remove_percents):\n",
+    "    ax0 = axarr[0, ix]\n",
+    "    ax1 = axarr[1, ix]\n",
+    "    rep = generate_rep_holdout(gene, cancer_type, remove_percent, method=projection_method)\n",
+    "    X_tr_train, X_tr_test, y_train, y_test = rep\n",
+    "    ax0.scatter(x=X_tr_train[y_train == 0, 0], y=X_tr_train[y_train == 0, 1], color='blue', label=z_label, s=16)\n",
+    "    ax0.scatter(x=X_tr_train[y_train == 1, 0], y=X_tr_train[y_train == 1, 1], color='orange', label=nz_label, s=16)\n",
+    "    ax0.set_xlabel(x_label)\n",
+    "    ax0.set_ylabel(y_label)\n",
+    "    ax0.set_title('Train data; holdout percent: {}'.format(remove_percent))\n",
+    "    ax1.scatter(x=X_tr_test[y_test == 0, 0], y=X_tr_test[y_test == 0, 1], color='blue', label=z_label, s=16)\n",
+    "    ax1.scatter(x=X_tr_test[y_test == 1, 0], y=X_tr_test[y_test == 1, 1], color='orange', label=nz_label, s=16)\n",
+    "    ax1.set_xlabel(x_label)\n",
+    "    ax1.set_ylabel(y_label)\n",
+    "    ax1.set_title('Test data; holdout percent: {}'.format(remove_percent))\n",
+    "\n",
+    "plt.tight_layout()\n",
+    "labels = [z_label, nz_label]\n",
+    "fig.suptitle('{} dimension reduction for varying holdout size: {}_{}'.format(\n",
+    "    projection_method.upper(), gene, cancer_type))\n",
+    "fig.legend(labels=labels, loc='lower center', ncol=2, fontsize=13)\n",
+    "fig.subplots_adjust(top=0.925, bottom=0.1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Running the above cell for a few different gene/cancer type combinations shows that in most of the cases where performance is fairly consistent across holdout data percentages (e.g. BRAF_THCA, TP53_LGG) the data is fairly close to being linearly separable. In other words, in these cases data appears to (roughly) cluster by mutation status.\n",
```

Comment:
Does separation in UMAP plots actually mean that two classes are linearly separable? Wouldn't it just mean that they're separable in the manifold space? 

I agree that separation = easier classification problem, I'm just asking about whether "linearly separable" is true here since I'm less familiar with the math than you.

Reply:
Also have you tried running the PCA plots? Do they look similar?

Reply:
> Does separation in UMAP plots actually mean that two classes are linearly separable? Wouldn't it just mean that they're separable in the manifold space?

Good question. I had been thinking of it as a supervised vs. unsupervised thing - my gut feeling was that if the labels separate fairly clearly when I apply unsupervised dimension reduction, then a supervised algorithm that's aware of the labels will almost certainly be able to separate them.

To address your point more specifically, I don't think there's any logical/mathematical reason that good UMAP separation in 2 dimensions necessarily implies good logistic regression hyperplane separation in 8000 dimensions - I feel like it would be possible to come up with a counterexample if you thought hard about it. But my suspicion is that this sort of thing would be uncommon in practical situations (although I haven't tested this).

I think if I was going to use this approach more generally to identify easy/hard classification problems I'd probably simulate some data to understand the tradeoffs a bit better, but for purposes of what I'm doing here I think it's probably fine to assume UMAP separation => decent classification in most cases (although I don't know if the reverse is true, even in practice).

> Also have you tried running the PCA plots? Do they look similar?

I have - for the most part they do look pretty similar. The "clusters" are a bit less obvious (higher variance), but the trends seem to be the same.

Reply:
This is a great point. Its been a while since looking at the umap paper, but my intuition is that if the labels were linearly separable, I'd expect to see a pretty clear separation in UMAP. If the labels were nonlinearly separable, I'd expect that you might see the separation in UMAP but not necessarily...I'd have to check how the math works though...just my intuition